scripts/link-vms-by-sr.py

```diff
@@ -46,61 +42,64 @@ def main(argv):
     for vm in vms:
         vmrec = vms[vm]
         # Ignore built-in templates
-        if vmrec['other_config'].has_key('default_template'):
-            if vmrec['other_config']['default_template'] == 'true':
-                continue
-        # Ignore dom0
-        if vmrec['is_control_domain']:
+        if vmrec["other_config"].get("default_template") == "true":
             continue
-        # Ignore snapshots
-        if vmrec['is_a_snapshot']:
+
+        # Ignore dom0 and Ignore snapshots
+        if vmrec["is_control_domain"] or vmrec["is_a_snapshot"]:
             continue
 
         # for each VM, figure out the set of SRs it uses
-        for vbd in vmrec['VBDs']:
-            if not vbds.has_key(vbd):
+        for vbd in vmrec["VBDs"]:
+            if vbd not in vbds:
                 continue
-            vdi = vbds[vbd]['VDI']
 
             # Ignore VBDs with no VDI such as an empty CD VBD
-            if vdi == '':
+            vdi = vbds[vbd]["VDI"]
+            if vdi == "" or vdi not in vdis:
                 continue
 
-            if not vdis.has_key(vdi):
+            sr = vdis[vdi]["SR"]
+            if sr not in srs:
                 continue
 
-            sr = vdis[vdi]['SR']
-            if not srs.has_key(sr):
-                continue
+            sruuid = srs[sr]["uuid"]
+            vmuuid = vmrec["uuid"]
+
+            vms_in_sr.setdefault(sruuid, {})[vmuuid] = 1
 
-            sruuid = srs[sr]['uuid']
-            vmuuid = vmrec['uuid']
-
-            if not vms_in_sr.has_key(sruuid):
-                vms_in_sr[sruuid] = {}
-            vms_in_sr[sruuid][vmuuid] = 1
-    
-    for sruuid in vms_in_sr.keys():
-        linkdir = "%s/by-sr/%s" % (dir, sruuid)
-        if os.path.isdir(linkdir):
-            print >> sys.stderr, "Directory %s already exists, skipping" % linkdir
+    return vms_in_sr
+
+
+def main():
+    session = XenAPI.xapi_local()
+    session.xenapi.login_with_password("", "", "1.0", "xen-api-scripts-linkvmsbysr.py")
+    atexit.register(logout, session)
+
+    input_dir = get_input_dir()
+    vms_in_sr = get_vms_in_sr(session)
+
+    for sruuid in list(vms_in_sr.keys()):
+        linkdir = "{}/by-sr/{}".format(input_dir, sruuid)
+        if Path(linkdir).is_dir():
+            print("Directory %s already exists, skipping" % linkdir, file=sys.stderr)
             continue
 
         try:
-            os.makedirs(linkdir)
+            Path(linkdir).mkdir(parents=True)
         except:
-            print >> sys.stderr, "Failed to create directory: %s" % linkdir
-        for vmuuid in vms_in_sr[sruuid].keys():
-            try:
-                src = "../../all/%s.vmmeta" % vmuuid
-                targ = "%s/%s.vmmeta" % (linkdir, vmuuid)
+            print("Failed to create directory: %s" % linkdir, file=sys.stderr)
+
+        try:
```

Comment:
You say this is for performance, but this changes behaviour, now if a symlink for a vm fails to be created can block the creation for the rest of the symlinks, this didn't happen before

Reply:
> You say this is for performance, but this changes behaviour, now if a symlink for a vm fails to be created can block the creation for the rest of the symlinks, this didn't happen before

Good catch!
I will revert this change.